lib/Controller/UsersController.php

```diff
@@ -469,6 +500,213 @@ public function create($username, $password, array $groups= [], $email='') {
 		);
 	}
 
+	/**
+	 * Set password for user using link
+	 *
+	 * @PublicPage
+	 * @NoCSRFRequired
+	 * @NoAdminRequired
+	 * @NoSubadminRequired
+	 *
+	 * @param $token
+	 * @param $userId
+	 * @return TemplateResponse
+	 */
+	public function setPasswordForm($token, $userId) {
+		try {
+			$this->checkPasswordSetToken($token, $userId);
+		} catch (UserTokenException $e) {
+			if ($e->getTokeFailureType() === UserTokenException::TOKEN_EXPIRED) {
+				return new TemplateResponse(
+					'user_management', 'new_user/resendtokenbymail', [
+						'link' => $this->urlGenerator->linkToRouteAbsolute('user_management.Users.resendToken', ['userId' => $userId]),
+						'userId' => $userId
+					], 'guest'
+				);
+			}
+			$this->log->error('Caught exception ' . $e->getMessage());
+			return new TemplateResponse(
+				'core', 'error', [
+				"errors" => [["error" => $this->l10n->t('Something went wrong. Please check the logs.')]]
+			],
+				'guest'
+			);
+		}
+
+		return new TemplateResponse(
+			'user_management', 'new_user/setpassword', [
+				'link' => $this->urlGenerator->linkToRouteAbsolute('user_management.Users.setPassword', ['userId' => $userId, 'token' => $token])
+			], 'guest'
+		);
+	}
+
+	/**
+	 * @param string $token
+	 * @param string $userId
+	 * @return null
+	 * @throws UserTokenException
+	 */
+	private function checkPasswordSetToken($token, $userId) {
+		$user = $this->userManager->get($userId);
+
+		$splittedToken = \explode(':', $this->config->getUserValue($userId, 'owncloud', 'lostpassword', null));
+		if (\count($splittedToken) !== 2) {
+			$this->config->deleteUserValue($userId, 'owncloud', 'lostpassword');
+			throw new UserTokenException(UserTokenException::TOKEN_INVALID);
```

Comment:
any better naming option? This seems misleading because what is invalid is what there is stored in the DB, not the token itself.

Reply:
Changed this to `TOKEN_AND_TIME_INVALID`. As token and time are stored in the db. Hope this looks ok?